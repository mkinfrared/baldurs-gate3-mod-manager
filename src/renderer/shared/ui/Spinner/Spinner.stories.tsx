import { Meta, StoryObj } from "@storybook/react";

import Text from "renderer/shared/ui/Text";

import Spinner from "./Spinner";
import css from "./Story.module.scss";

type Story = StoryObj<typeof meta>;

const meta = {
  title: "UI/Spinner",
  component: Spinner,
  tags: ["autodocs"],
} satisfies Meta<typeof Spinner>;

const spinnerTypes = [
  "bars",
  "oval",
  "puff",
  "rings",
  "ripple",
  "tail-spin",
] as const;

const Default: Story = {
  render: () => (
    <div className={css.Story}>
      {spinnerTypes.map((spinnerType) => (
        <div key={spinnerType} className={css.loader}>
          <Text>{spinnerType}</Text>
          <Spinner type={spinnerType} />
        </div>
      ))}
    </div>
  ),
};

export { Default };

export default meta;
