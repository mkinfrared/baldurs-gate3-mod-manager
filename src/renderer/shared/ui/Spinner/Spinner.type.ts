export type SpinnerProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  type?: "bars" | "oval" | "puff" | "rings" | "ripple" | "tail-spin";
};
