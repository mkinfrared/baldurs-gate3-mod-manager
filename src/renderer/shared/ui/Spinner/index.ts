export { default } from "./Spinner";

export * from "./Spinner.type";
