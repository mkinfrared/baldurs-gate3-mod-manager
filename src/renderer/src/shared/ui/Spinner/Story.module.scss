.Story {
  display: grid;
  grid-auto-rows: 200px;
  grid-template-columns: repeat(4, 200px); /* stylelint-disable-line */

  .loader {
    align-items: center;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 20px;

    p {
      color: #ffffff;
      margin-bottom: 10px;
    }

    svg {
      height: 100%;
      width: 100%;
    }
  }
}
