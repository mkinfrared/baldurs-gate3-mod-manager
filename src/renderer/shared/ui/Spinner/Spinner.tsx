import Bars from "renderer/shared/assets/spinners/bars.svg";
import Oval from "renderer/shared/assets/spinners/oval.svg";
import Puff from "renderer/shared/assets/spinners/puff.svg";
import Rings from "renderer/shared/assets/spinners/rings.svg";
import Ripple from "renderer/shared/assets/spinners/ripple-1s-200px.svg";
import TailSpin from "renderer/shared/assets/spinners/tail-spin.svg";
import { classNames } from "renderer/shared/lib/helpers";

import css from "./Spinner.module.scss";
import { SpinnerProps } from "./Spinner.type";

const Spinner = ({ className, type = "oval" }: SpinnerProps) => {
  const props = { className: classNames(css.Spinner, className) };

  switch (type) {
    case "bars":
      return <Bars {...props} />;
    case "oval":
      return <Oval {...props} />;
    case "puff":
      return <Puff {...props} />;
    case "rings":
      return <Rings {...props} />;
    case "ripple":
      return <Ripple {...props} />;
    case "tail-spin":
      return <TailSpin {...props} />;
    default:
      return <Oval {...props} />;
  }
};

export default Spinner;
