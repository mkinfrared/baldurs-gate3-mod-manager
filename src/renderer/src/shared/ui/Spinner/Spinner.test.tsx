import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import Spinner from "./Spinner";

describe("<Spinner />", () => {
  const Component = <Spinner type="oval" />;

  it("should be defined", () => {
    expect(Spinner).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should return a bars loader", () => {
    const type = "bars";
    const { getByText } = render(<Spinner type={type} />);

    expect(getByText(/bars/)).toBeDefined();
  });

  it("should return an oval loader", () => {
    const type = "oval";
    const { getByText } = render(<Spinner type={type} />);

    expect(getByText(/oval/)).toBeDefined();
  });

  it("should return a puff loader", () => {
    const type = "puff";
    const { getByText } = render(<Spinner type={type} />);

    expect(getByText(/puff/)).toBeDefined();
  });

  it("should return a rings loader", () => {
    const type = "rings";
    const { getByText } = render(<Spinner type={type} />);

    expect(getByText(/rings/)).toBeDefined();
  });

  it("should return a ripple loader", () => {
    const type = "ripple";
    const { getByText } = render(<Spinner type={type} />);

    expect(getByText(/ripple/)).toBeDefined();
  });

  it("should return a tail-spin loader", () => {
    const type = "tail-spin";
    const { getByText } = render(<Spinner type={type} />);

    expect(getByText(/tail-spin/)).toBeDefined();
  });

  it("should return an oval loader by default", () => {
    const { getByText } = render(<Spinner />);

    expect(getByText(/oval/)).toBeDefined();
  });
});
