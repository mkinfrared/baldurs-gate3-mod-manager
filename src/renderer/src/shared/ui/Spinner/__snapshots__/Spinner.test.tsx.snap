// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Spinner /> > should match the snapshot 1`] = `
<div>
  <svg
    class="Spinner"
  >
    oval.svg
  </svg>
</div>
`;
